### PR TITLE
lua: do not set metatable for arrays/maps

### LINKED ISSUE #9655: Lua: utils.format_json omits all fields from a table created by utils.parse_json('[]') without error indication
### Important Information
mpv version 0.33.1 on Ubuntu 21.10 amd64

There is a bug in either `utils.parse_json` or `utils.format_json` that can cause the latter to return an empty array `[]` but no error code (`nil`), when `utils.parse_json` is used on a JSON structure containing an empty table that is later populated.

Making a deep copy of the `utils.parse_json` return value works around the problem. So it appears to be some hidden property of the table returned by `utils.parse_json` that causes the resulting data loss in `utils.format_json`.

I understand that empty tables are a touchy issue in Lua vs. JSON, but `utils.format_json` should at least return an error instead of silently omitting all table entries if `utils.parse_json` can't be fixed to return "normal" empty tables from `[]`.

I use JSON to store a "database" for some metadata values and noticed some fields would just disappear on saving until I learned to make a deep copy of the structure before conversion to JSON. Within Lua itself the table works normally and the fields keep their values however.

### Reproduction steps
Minimal test.lua script, first converting a structure with empty table to JSON, then reconstructing and populating it but it remains empty with `utils.format_json`:
```lua
local utils=require 'mp.utils'

local tbl={data={}}
local json,err=utils.format_json(tbl)
print(json,err)
-- OK? '{"data":[]}',nil

tbl=utils.parse_json(json)
tbl.data["field"]=123
print(utils.to_string(tbl))
-- OK, field is set: '{"data" = {"field" = 123}}'

print(utils.format_json(tbl))
-- WRONG: '{"data":[]}',nil
-- Expected: either '{"data":{"field":123}}',nil or nil,'error'
```

### Expected behavior
`utils.format_json` should either return the correct JSON string with all table entries, or an error, so that JSON can be used to store and recover Lua tables.

### Actual behavior
`utils.format_json` returns an empty array `[]` for the table, but the error returned is nil.

### Log file
```
[test] {"data":[]} nil 
[test] {"data" = {"field" = 123}} 
[test] {"data":[]} nil 
```

### Sample files

see above

### PR DESCRIPTION
**Pull request by TheAMM**
```
Lua has only tables, which function as arrays and maps. In makenode(), mpv has
to probe the tables and decide whether it is an MPV_FORMAT_ARRAY or MPV_FORMAT_MAP.
In case of an empty table, array is assumed. Alternatively, if the table has
been assigned one of mpv's format metatables, as in pushnode(), the probe can
be skipped.

A bug appears. If the user modifies tables mpv has returned in a manner that
doesn't fit the assigned array/map and passes them back to mpv, an array-table
will have its non-index keys ignored. Map-tables are not particularly affected,
since non-string map keys are already illegal and raise an error.
Easy way to trip into this is to serialize an empty table with utils.format_json()
into assumed "[]", and later parsing that into a table. The table is now tagged as
an array, and modifications to its fields are ignored if again serialized.

This commit removes using metatables for checking and setting array/map
table type in makenode() and pushnode(), The UNKNOWN_TYPE metatable is kept.

Performance impact is trivial, especially for the common case where mpv is given
tables created on the Lua side, with no metatable set.
For extreme cases, like json-serializing a table with 10k tables with 100 numbers
each, using the metatables has allowed for about 4% speedup, although only for
data that has arrived from mpv and been metatabled first.

Fixes #9655
```

Diff looks a bit dirty in makenode(), but I just remove the metatable checking block and [dedent](https://github.com/mpv-player/mpv/blob/c9c40d53863f41b58eb8bfbb8a43bef02612b810/player/lua.c#L698-L722) the [`if (format == MPV_FORMAT_NONE) {` block](https://github.com/mpv-player/mpv/blob/f40d66f62454b66280a5a25c3f718fdafba860cf/player/lua.c#L709-L735), which is now always ran.

### CODE REVIEW

**Comment by avih** [comment]
> Performance impact is trivial, especially for the common case where mpv is given
tables created on the Lua side, with no metatable set.

In this case there's no performance impact (well, there is - it should improve).

The only relevant test case is that an mpv-node is converted to a lua table, and then this table is converted back to mpv-node, and that's the only case which needs performance numbers.

**Comment by avih** [comment]
I just looked at the commit which added this tag system - 15134935602a546103a26f4b3ecc72da25896f09 and it seems the reason was not performance, but rather specifically to be able to distinguish empty array from empty map when such table is created from an mpv-node and later converted back into mpv-node.

So I think this system should stay, because it does have value, but only applied when the table is empty.

The performance improvement of this system is a side effect, and it's indeed small, so we can live without it.

**Comment by avih** [comment]
On further thought, this tagging system is useful, and I think we should keep it - and document it.

It's useful because without it, for ambiguous tables we have to guess, but with the tagging, we can identify unambiguously at least tables which were originally from an mpv-node.

Even more, I think we should allow scripts to tag a table as array or map, so that when it's sent to mpv (to `set_property_native`, or `format_json` etc) - it can reflect the intent without ambiguity.

Without the tagging, if a user has an ambigous table (with regards to JSON or mpv-node), then they have no way to specify how it should be encoded.

Do note that not only empty tables are ambiguous. Any array is also ambiguous, because it could be either an array normally (JSON `["foo", "bar", "baz"]`) or a map (JSON `{"1":"foo", "2":"bar", "3":"baz"}`), and the user currently can't specify which type it should be considered.

Bottom line, I think we should:
- Keep the tag system.
- Document that all tables from an mpv API are tagged as array/map.
- Allow scripts to tag tables as array or map (can also be used to change its tag).
- Unrelated to the above, but with the same goal of allowing to represent more things as mpv-node: allow array with holes/nil in the middle (like JSON and mpv-node allow). TBD.
